Src/Web/Web.Shared.Net/RequestTrackingTelemetryModule.cs

```diff
@@ -347,5 +382,151 @@ private bool TryInitializeCorrelationHelperIfNotInitialized()
+            /// </summary>
+            internal int MAXSIZE { get; private set; }
+
+            /// <summary>
+            /// Request will be tagged with an id to identify if it should be logged later.
+            /// </summary>
+            internal void OnBeginRequest_IdRequest(HttpContext context)
+            {
+                this.TagRequest(context);
+            }
+
+            /// <summary>
+            /// OnEndRequest - Should this request be logged?
+            /// Will compare a request id against a hash-set of known requests.
+            /// If this request is not known, add it to hash-set and return true (safe to log).
+            /// If this request is known, return false (do not log twice).
+            /// Additional requests with the same id will return false.
+            /// </summary>
+            internal bool OnEndRequest_ShouldLog(HttpContext context)
+            {
+                var headers = context.Request.Headers;
+
+                var rootRequestId = headers[HeaderRootRequestId];
+                if (rootRequestId != null)
+                {
+                    if (!this.IsRequestKnown(rootRequestId))
+                    {
+                        // doesn't exist add to dictionary and return true
+                        this.AddRequestToDictionary(rootRequestId);
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            /// <summary>
+            /// Tag new requests.
+            /// Transfer Ids to parent requests.
+            /// </summary>
+            private void TagRequest(HttpContext context)
+            {
+                var newId = Guid.NewGuid().ToString();
+
+                var headers = context.Request.Headers;
+
+                if (headers[HeaderRootRequestId] == null)
+                {
+                    headers[HeaderRootRequestId] = newId;
+                }
+
+#if DEBUG
+                // additional ids help developer watch request hierarchy while debugging.
+                if (headers[HeaderRequestId] != null)
+                {
+                    headers[HeaderParentRequestId] = headers[HeaderRequestId];
+                }
+
+                headers[HeaderRequestId] = newId;
+#endif
+            }
+
+            /// <summary>
+            /// Has this request been tracked.
+            /// </summary>
+            private bool IsRequestKnown(string requestId)
+            {
+                return activeRequestsA.ContainsKey(requestId) || activeRequestsB.ContainsKey(requestId);
+            }
+
+            /// <summary>
+            /// Track this requestId.
+            /// </summary>
+            /// <remarks>
+            /// Dictionary A will be read/write.
+            /// When dictionary A is full, move to B and create new A.
+            /// Dictionary B will be read-only.
+            /// </remarks>
+            private void AddRequestToDictionary(string requestId)
```

Comment:
We were discussing with @SergeyKanzhelev offline on how to limit the lookup time on this dictionary when it grows big enough to have hash collisions but not yet `MAXSIZE` and also on how to limit memory impact of the current solution (let's consider these suggestions in the different enhancement with lower priority):
- We can add `remove()` back but without the complex logic to handle all the special cases by sacrificing some of those;
- We can store Base64 of GUID string in the dictionary or any other non-string value that will be more memory-friendly (e.g. GUID object itself);